src/Groups/MatGrp.jl

```diff
@@ -0,0 +1,790 @@
+import AbstractAlgebra: MatElem, matrix, MatSpace, parent_type, Ring, RingElem
+import Hecke: base_ring, det, fmpz, fq_nmod, FqNmodFiniteField, nrows, tr, trace
+import GAP: FFE
+
+export
+    general_linear_group,
+    mat_elem_type,
+    matrix_group,
+    MatrixGroup,
+    MatrixGroupElem,
+    omega_group,
+    orthogonal_group,
+    ring_elem_type,
+    special_linear_group,
+    special_orthogonal_group,
+    special_unitary_group,
+    symplectic_group,
+    unitary_group,
+    GL, GO, GU, SL, SO, Sp, SU
+
+
+
+abstract type AbstractMatrixGroupElem <: GAPGroupElem{GAPGroup} end
+
+# NOTE: always defined are deg, ring and at least one between { X, gens, descr }
+# NOTE: the field mat_iso are always defined if the field X is
+"""
+    MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
+Type of groups `G` of `n x n` matrices over the ring `R`, where `n = degree(G)` and `R = base_ring(G)`.
+
+At the moment, only rings of type `FqNmodFiniteField` are supported.
+"""
+mutable struct MatrixGroup{RE<:RingElem, T<:MatElem{RE}} <: GAPGroup
+   deg::Int
+   ring::Ring
+   X::GapObj
+   gens::Vector{<:AbstractMatrixGroupElem}
+   descr::Symbol                       # e.g. GL, SL, symbols for isometry groups
+   mat_iso::GenMatIso
+   order::fmpz
+   AbstractAlgebra.@declare_other
+
+   MatrixGroup(m::Int, F::Ring) = new{elem_type(typeof(F)), dense_matrix_type(elem_type(typeof(F)))}(m,F)
+
+end
+
+MatrixGroup{RE,T}(m::Int, F::Ring) where {RE,T} = MatrixGroup(m,F)
+
+# build a MatrixGroup given a list of generators, given as array of either MatrixGroupElem or AbstractAlgebra matrices
+# WARNING: if the elements of V have type MatElem, it does not check whether the determinant is nonzero
+function MatrixGroup(m::Int, F::Ring, V::AbstractVector{T}) where T<:Union{MatElem,AbstractMatrixGroupElem}
+   G = MatrixGroup(m,F)
+
+   L = Vector{MatrixGroupElem}(undef, length(V))
+   for i in 1:length(V)
+      @assert base_ring(V[i])==F "The elements must have the same base ring"
+      @assert nrows(V[i])==m "The elements must have the same dimension"
+      if T<:MatElem
+         L[i] = MatrixGroupElem(G,V[i])
+      else
+# TODO: this part of code can be replaced by the following line
+# L[i] = deepcopy(V[i]); L[i].parent = G;
+         if isdefined(V[i],:elm)
+            if isdefined(V[i],:X)
+               L[i] = MatrixGroupElem(G,V[i].elm,V[i].X)
+            else
+               L[i] = MatrixGroupElem(G,V[i].elm)
+            end
+         else
+            L[i] = MatrixGroupElem(G,V[i].X)
+         end
+# once deepcopy works for GAP objects
```

Comment:
Maybe this line should be moved upwards, to lines 61/62? I was really confused by lines 61/62 (as in: "so why not do it?") until I saw this line